### PR TITLE
Cleanup and features

### PR DESCRIPTION
**Pull request by RuthgerD**
code is self documenting :^)

### CODE REVIEW

**Comment by RuthgerD** [comment]
hope it merged properly

File: src/main/java/org/coffeemine/app/spring/auth/BasicAccessControl.java [hunk 1]
```diff
@@ -48,7 +48,7 @@ public int getUserId() {
     public void signOut() {
         userId = null;
         CurrentUser.set(null);
-        UI.getCurrent().getPage().reload();
+        UI.getCurrent().navigate(LoginScreen.class);
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/auth/BasicAccessControl.java]: Nice

File: src/main/java/org/coffeemine/app/spring/components/AddProjectDialog.java [hunk 2]
```diff
@@ -4,31 +4,64 @@
...
+        projectDescription.setWidthFull();
+        projectDescription.getStyle().set("margin", "0px");
+
+        addCleanButton.addClickListener(c -> {
+            try {
+                final var newProjectName = projectName.getOptionalValue().orElseThrow();
+                NitriteDBProvider.getInstance().addProject(new Project(newProjectName, 0));
+                Notification.show("Project \"" + newProjectName+ "\" created", 2000, Notification.Position.BOTTOM_END);
+                close();
+            } catch (Exception e) {}
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/AddProjectDialog.java]: Exceptions for control flow is no thanks
I love it for the memes, but it's definitely a bad idea in all languages where stack unwinding is heavy

File: src/main/java/org/coffeemine/app/spring/components/HoursLogging.java [hunk 3]
```diff
@@ -26,16 +34,42 @@ public HoursLogging() {
             fragment.setBegin(start_time.getValue());
             fragment.setEnd(end_time.getValue());
             // TODO: save selected task
+            final var id = db.addFragment(fragment);
+            task.addFragment(id);
+            Notification.show(
+                    "Task is now added!",
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/HoursLogging.java]: Not a task, but a fragment. And anyway it should just show "sucess" or something like that

File: src/main/java/org/coffeemine/app/spring/components/HoursLogging.java [hunk 4]
```diff
@@ -6,14 +6,22 @@
...
 import java.time.LocalDate;
+import java.util.stream.Collectors;
 
 public class HoursLogging extends Dialog {
 
+    private ITask task = null;
     public HoursLogging() {
 
         // TODO: show Sprint selector
         // TODO: show Task selector (only show tasks for selected sprint)
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/HoursLogging.java]: Shouldn't this be gone now?

File: src/main/java/org/coffeemine/app/spring/components/ProjectList.java [hunk 5]
```diff
@@ -46,10 +51,24 @@ private void generate(Modes mode, int limit) {
...
+                        UI.getCurrent().navigate("");
+                        redirectDialog.close();
+                    });
+                    confirm.addThemeVariants(ButtonVariant.MATERIAL_CONTAINED);
+                    final var cancel = new Button("Cancel", c -> redirectDialog.close());
+                    cancel.getStyle().set("margin-left", "auto");
+                    buttons.add(cancel, confirm);
+                    redirectDialog.add(new H3("Switch the current project to: \"" + project.getName() + "\"?"), buttons);
+
+                    redirectDialog.open();
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/ProjectList.java]: Should only ask if this isn't the currently selected project

File: src/main/java/org/coffeemine/app/spring/data/Sprint.java [hunk 6]
```diff
@@ -5,11 +5,13 @@
...
 
 import javax.validation.constraints.NotNull;
+
+import java.time.Instant;
 import java.time.LocalDate;
 import java.time.format.DateTimeFormatter;
 import java.util.ArrayList;
 
-public class Sprint implements ISprint {
+public class Sprint implements ISprint, ChangeTracker {
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/data/Sprint.java]: `ISprint` should extend `ChangeTracker` directly

File: src/main/java/org/coffeemine/app/spring/components/RecentChangesView.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        db.getSprints4Project(project).forEach(sprint -> {
+            total = Stream.concat(total, db.getTasks4Sprint(sprint).map(task -> (ChangeTracker) task));
+        });
+
+        total.sorted((a, b) -> -Long.compare(a.getLastModifiedTime(), b.getLastModifiedTime()))
+        .forEach(change -> {
+            final var date = LocalDateTime.ofInstant(Instant.ofEpochMilli(change.getLastModifiedTime()), ZoneId.systemDefault());
+
+            final var changeText = new Span();
+            changeText.getElement().setProperty("innerHTML", change.getType() + " \"" + change.getMessage() + "\":<br/>"
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/RecentChangesView.java]: Looks fun

File: src/main/java/org/coffeemine/app/spring/components/RecentChangesView.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    public RecentChangesView(Project project) {
+        addClassName("recent-changes");
+        add(new H2("Recent Changes:"));
+
+        final var db = NitriteDBProvider.getInstance();
+
+        total = Stream.concat(total, db.getSprints4Project(project).map(sprint -> (ChangeTracker) sprint));
+
+        db.getSprints4Project(project).forEach(sprint -> {
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/components/RecentChangesView.java]: Are you sure this isn't a place for `.flatMap` ?

File: src/main/java/org/coffeemine/app/spring/data/Task.java [hunk 9]
```diff
@@ -5,9 +5,11 @@
 import org.dizitart.no2.Document;
 
 import javax.validation.constraints.NotNull;
+
+import java.time.Instant;
 import java.util.ArrayList;
 
-public class Task implements ITask {
+public class Task implements ITask, ChangeTracker {
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/data/Task.java]: `ITask` should extend `ChangeTracker`

File: src/main/java/org/coffeemine/app/spring/view/TasksView.java [hunk 10]
```diff
@@ -10,32 +14,40 @@
...
-        removeAll();
-        var db = NitriteDBProvider.getInstance();
-
-        db.getTasks4Sprint(db.getSprint(localSprint)).forEach(task -> {
-            var assigneeText = "";
-
-            for (var user : task.getAssignees()) {
-                assigneeText += db.getUser(user).getName();
+        if (sprint != null) {
+            var db = NitriteDBProvider.getInstance();
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/view/TasksView.java]: How come this one isn't final ?

File: src/main/java/org/coffeemine/app/spring/view/TasksView.java [hunk 11]
```diff
@@ -10,32 +14,40 @@
...
+
+            db.getTasks4Sprint(db.getSprint(sprint.getId())).forEach(task -> {
+                var assigneeText = "";
+    
+                for (var user : task.getAssignees()) {
+                    assigneeText += db.getUser(user).getName();
+                }
```
Comment (AeroStun) [inline src/main/java/org/coffeemine/app/spring/view/TasksView.java]: `for` in `.forEach` means not enough `Stream`ing action

File: src/main/java/org/coffeemine/app/spring/components/RecentChangesView.java (hunk 7 above)
Comment (RuthgerD) [reply src/main/java/org/coffeemine/app/spring/components/RecentChangesView.java]: normal \n line break doesn't get respected so I wont be respecting vaadin :+1:

File: src/main/java/org/coffeemine/app/spring/components/AddProjectDialog.java (hunk 2 above)
Comment (RuthgerD) [reply src/main/java/org/coffeemine/app/spring/components/AddProjectDialog.java]: hmm, my thinking was that this avoids a bunch of if statements

**Pull request merged**